epsagon/lambda_trigger.go

```diff
@@ -43,50 +35,106 @@ func mapParametersToString(params map[string]string) string {
-	event, ok := rawEvent.(lambdaEvents.APIGatewayProxyRequest)
-	if !ok {
-		tracer.AddException(&protocol.Exception{
-			Type: "trigger-creation",
-			Message: fmt.Sprintf(
-				"failed to convert rawEvent to lambdaEvents.APIGatewayProxyRequest %v",
-				rawEvent),
-			Time: tracer.GetTimestamp(),
-		})
-		return nil
+type APIGatewayEventFields struct {
+	requestId             string
+	headers               map[string]string
+	host                  string
+	httpMethod            string
+	stage                 string
+	queryStringParameters map[string]string
+	pathParameters        map[string]string
+	path                  string
+	body                  string
+}
+
+func getEventAssertionException(rawEvent interface{}, assertedType string) *protocol.Exception {
+	return &protocol.Exception{
+		Type:    "trigger-creation",
+		Message: fmt.Sprintf("failed to convert rawEvent to %s. %v", assertedType, rawEvent),
+		Time:    tracer.GetTimestamp(),
 	}
-	triggerEvent := &protocol.Event{
-		Id:        event.RequestContext.RequestID,
+}
+
+func getAPIGatewayTriggerEvent(eventFields *APIGatewayEventFields, metadataOnly bool) *protocol.Event {
+	triggerEvent := getAPIGatewayBaseEvent(eventFields)
+	if !metadataOnly {
+		addAPIGatewayEventMetadata(triggerEvent, eventFields)
+	}
+	return triggerEvent
+}
+
+func getAPIGatewayBaseEvent(eventFields *APIGatewayEventFields) *protocol.Event {
+	return &protocol.Event{
+		Id:        eventFields.requestId,
 		Origin:    "trigger",
 		StartTime: tracer.GetTimestamp(),
 		Resource: &protocol.Resource{
-			Name:      event.Headers["Host"],
+			Name:      eventFields.host,
 			Type:      "api_gateway",
-			Operation: event.HTTPMethod,
+			Operation: eventFields.httpMethod,
 			Metadata: map[string]string{
-				"stage":                   event.RequestContext.Stage,
-				"query_string_parameters": mapParametersToString(event.QueryStringParameters),
-				"path_parameters":         mapParametersToString(event.PathParameters),
-				"path":                    event.Resource,
+				"stage":                   eventFields.stage,
+				"query_string_parameters": mapParametersToString(eventFields.queryStringParameters),
+				"path_parameters":         mapParametersToString(eventFields.pathParameters),
+				"path":                    eventFields.path,
 			},
 		},
 	}
-	if !metadataOnly {
-		if bodyJSON, err := json.Marshal(event.Body); err != nil {
-			tracer.AddException(&protocol.Exception{
-				Type:      "trigger-creation",
-				Message:   fmt.Sprintf("Failed to serialize body %s", event.Body),
-				Traceback: string(debug.Stack()),
-				Time:      tracer.GetTimestamp(),
-			})
-			triggerEvent.Resource.Metadata["body"] = ""
-		} else {
-			triggerEvent.Resource.Metadata["body"] = string(bodyJSON)
-		}
-		triggerEvent.Resource.Metadata["headers"] = mapParametersToString(event.Headers)
+}
+
+func addAPIGatewayEventMetadata(triggerEvent *protocol.Event, eventFields *APIGatewayEventFields) {
```

Comment:
would rename the func name addXXXXPayload/non metadata (as its being called if !metadataOnly)

Reply:
Good idea, done.